strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java

```diff
@@ -197,16 +269,70 @@ public void download(@ApiParam(value = "The storageId", required = true)
 
             response.setStatus(NOT_FOUND.value());
 
-            return;
+            return null;
         }
+        
+        return is;
+    }
+    
+    private RepositoryFileAttributes getRepositoryFileAttributes(String storageId,
+                                                                 String repositoryId,
+                                                                 String path,
+                                                                 HttpServletResponse response) 
+                                       throws NoSuchAlgorithmException,
+                                              ProviderImplementationException
+    {
+        RepositoryFileAttributes fileAttributes;
+        try
+        {
+            fileAttributes =  getArtifactManagementService().getAttributes(storageId, repositoryId, path);
 
-        setMediaTypeHeader(path, response);
-
-        response.setHeader("Accept-Ranges", "bytes");
+            if (fileAttributes == null)
+            {
+                response.setStatus(NOT_FOUND.value());
+                return null;
+            }
+        }
+        catch (ArtifactTransportException e)
+        {
+            logger.debug("Unable to find artifact by path " + path, e);
 
-        ArtifactControllerHelper.setHeadersForChecksums(is, response);
+            response.setStatus(NOT_FOUND.value());
 
-        logger.debug("Download succeeded.");
+            return null;
+        }
+        
+        return fileAttributes;
+    }
+    
+    
+    private void setHeaders(ArtifactInputStream ais,
+                            HttpServletResponse response,
+                            RepositoryFileAttributes fileAttributes,
+                            String path) 
+                 throws IOException
+    {      
+        response.setHeader("Accept-Ranges", "bytes");
+        
+        if(!response.containsHeader("Content-Length"))
+        {
+            long totalBytes = 0L;
+            int readLength;
+            byte[] bytes = new byte[4096];
+            
+            while ((readLength = ais.read(bytes, 0, bytes.length)) != -1)
+            {
+                totalBytes += readLength;
+            }
+        
+            response.setHeader("Content-Length", Long.toString(totalBytes));
```

Comment:
@carlspring I don't think counting input stream bytes just for displaying result as a `Content-Length` header is a good approach here.

Reply:
I had not better idea about guessing the size. I mean... once you're done writing to the stream, you'll have the final number of bytes.

Do you know of a better way?

Reply:
We have 

`org.carlspring.strongbox.domain.ArtifactEntry#sizeInBytes` or `org.carlspring.strongbox.providers.io.RepositoryFileAttributes#size`